Add a class to help manage and interpolate static datasets

MiCurry opened the pull request:
This merge adds a new module to the init_atmosphere core, mpas_geotile_manager, which contains a class, mpas_geotile_mgr_type, which can be used to help read, store and interpolate static fields. Currently, it is only being compiled, but in the future will be used to parallelize the interpolation of static fields.

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1000 @@
+        ! 0:tile_nx, and 0:tile_ny).
+        !
+        x = (start_x - 1) / tile_nx
+        if (x < 0 .or. y > size(mgr % hash, 1)) then
```
As pointed out by the gfortran compiler, it seems that `y` is indeed being used before it has been defined here.
```
mpas_geotile_manager.F:373:0:                                                                                                        
                                                                                                                                     
  373 |         if (x < 0 .or. y > size(mgr % hash, 1)) then                                                                         
      |                                                                                                                              
Warning: ‘y’ may be used uninitialized in this function [-Wmaybe-uninitialized]
```

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1000 @@
+        ! Left
+        neighbor => null()
+        if (tile % hash_x -1 < 0) then
+            xHash = nTileX - 1
```
The gfortran compiler seems correct in identifying `nTileX` as being used here before it has been defined:
```
mpas_geotile_manager.F:847:0:                                                                                                        
                                                                                                                                     
  847 |             xHash = nTileX - 1                                                                                               
      |                                                                                                                              
Warning: ‘ntilex’ may be used uninitialized in this function [-Wmaybe-uninitialized]
```

MiCurry replied:
nTileX and nTileY were both actually being defined here in the deceleration section, but not used as they are being replaced by `mgr % nTileX` and `mgr % ntileY`.

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1000 @@
+        ! NOTE: This calculation assumes that a dataset is a global dataset and may
+        ! not work correctly for non-global datasets
+        mgr % pixel_nx = 360.0_RKIND / dx 
+        mgr % pixel_ny = 180.0_RKIND / dy
```
When computing `mgr % pixel_nx` and `mgr % pixel_ny`, I think we should be careful to round to the nearest integer rather than truncating to an integer. For example, a 15-arc-second dataset would have dx ~= 0.00416666667, which would give a `mgr % pixel_nx` value of 86399 with the code as written, when the correct value is 86400.

MiCurry commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1000 @@
+    !> will be in the range of -1/2 * pi to 1/2 * pi and 0 to 2.0 * pi, respectively.
+    !
+    !-----------------------------------------------------------------------
+    subroutine mpas_geotile_mgr_tile_to_latlon(mgr, tile, j, i, lat, lon)
```
When testing this function, I was a bit confused by the range of latitudes that it produced. If I grabbed the most north east tile `mgr % get_tile(5. * pi, 2.0 * pi, tile)` some of the range of latitudes produced by `tile_to_latlon` were greater than .5 * pi, and I'm not sure exactly why that would be.

A first thought is that this is because `known_lat` is not equal to exactly 5. * -pi. @mgduda, is that correct?

mgduda replied:
Typically, the `known_lat` and `known_lon` values for a dataset give the coordinates of the *center* of a pixel. So if the boundary of the pixel extends to, e.g., the South Pole, then the pixel's center coordinates will be greater than -Pi/2.

mgduda replied:
I just left a comment for lines 642 and 643 that may be related.

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1000 @@
+            xHash = modulo(tile % hash_x + (mgr % nTileX / 2), mgr % nTileX)
+            yHash = tile % hash_y
+        else
+            xHash = tile % hash_x + 1
```
Why is 1 being added to `tile % hash_x` here?

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1000 @@
+        neighbor => null()
+        if (tile % hash_x + 1 > mgr % nTileX - 1) then
+            yHash = 1
+            xHash = tile % hash_y
```
I don't quite understand the logic for setting `yHash` and `xHash` here.

MiCurry commented:
@mgduda thanks for the primarily review. I just pushed an update fixing the warnings and other issues you found.

MiCurry commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,998 @@
+        ! Check to see if the directory exists
+        inquire(file=trim(path), exist=res)
+        if (.not. res) then
+            call mpas_log_write("Could not find geotile directory: "//path, messageType=MPAS_LOG_ERR)
```
`path` needs to be `trim(path)` here.

MiCurry commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,998 @@
+        ! Check to see if the index file exists in the directory
+        inquire(file=trim(mgr % directory)//"index", exist=res)
+        if (.not. res) then
+            call mpas_log_write("Could not find an 'index' file in geotile directory: "//path, messageType=MPAS_LOG_ERR)
```
Aswell, `path` needs to be `trim(path)` here.

MiCurry commented:
@mgduda, I have just pushed an update to this branch that includes the changes above. I have pulled them out of 469c271 and moved them into 9111cfb and I have added the `convert_lx` routine in fe0709a.

MiCurry commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1025 @@
+        ierr = 0
+
+        ! Check to see if the directory exists
+        inquire(file=trim(path), exist=res)
```
It seems to appear that ifort will set `res` to `.false.` on the above `inquire` call, but gfortran does not.

mgduda replied:
This check may not really be necessary. The check in the next code block on whether the `index` file exists is a sufficient condition for the directory existing.

MiCurry replied:
Actually, taking a look at the standard, it doesn't appear that `inquire` is able to work on directories. From section 9.9 of the Fortran 2003 standard:

> The INQUIRE statement may be used to inquire about properties of a particular named file or of the
> 25 connection to a particular unit.

And then in 9.4:

> A unit, specified by an io-unit, provides a means for referring to a file.

So it looks like gfortran's ability to inquire on a directory is an extension, similarly to ifort's `directory` keyword.


I also don't think this `inquire` statement is really needed, as the next `inquire` (into the index file) will return `.false.` if the directory doesn't exist. So I think its safe to remove it.

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1025 @@
+        call rad2Deg(lon)
+
+        pixel_x = int((lon - known_lon) / dx)
+        pixel_y = int((lat - known_lat) / dy)
```
In lines 692 and 693, I think `int` should be `nint`.

MiCurry replied:
Trying out `nint` (as well as `floor` and `ceiling`) will cause the geotile_manager to generate tile names that don't exist for the dataset. Although, I'm not sure if this behavior is correct or not. For instance the geotile_manager will generate the tile `42001-43200.21601-22800` for the point: ```(1.57080, 0.00000)```. The calculation is:

```
* lat = 90.0
* (lat - known_lat) / dx = 21599.5059
* nint(21599.5059) = 21600
```

However, I don't think this is correct, as the very last tile contains the `21600` pixel: `42001-43200.20401-21600`. 

So perhaps the tile filename generation is incorrect? Perhaps off by one?

mgduda replied:
Are the pixel coordinates that are returned by `mpas_geotile_mgr_latlon_to_pixel` supposed to be 0-based or 1-based?

mgduda replied:
Let's assume that every latitude value in the interval [-Pi/2, Pi/2] is uniquely mapped to a j-index. Then for the center latitude, *lat*, of a given pixel, we can either say that all latitudes in the interval [*lat* - Pi/2, *lat* + Pi/2) map to that pixel, or that all latitudes in the interval (*lat* - Pi/2, *lat* + Pi/2] map to that pixel. Regardless of which we choose, there will always be a pole, i.e., -Pi/2 or Pi/2 that is not mapped to some j-index and that therefore needs special treatment.

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1025 @@
+        call mpas_pool_get_config(mgr % pool, 'dy', dy)
+
+        lat = real((j - (tile_bdr + 1) + tile % y - 1), kind=RKIND) * dy + known_lat + 1.0
+        lon = real((i - (tile_bdr + 1) + tile % x - 1), kind=RKIND) * dx + known_lon + 1.0
```
Where does the `+ 1.0` at the end of the expressions for `lat` and `lon` come from?

MiCurry replied:
I'm not sure where it comes from. Previous implementations didn't have it. On its own, it doesn't make much sense to add 1.0 to the latitude or longitude.

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1025 @@
+            loaded = .true.
+        endif
+
+    end function
```
To match the prevailing style in MPAS, we could change this line to
```fortran
    end function mpas_geotile_mgr_search_tile
```

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1025 @@
+        !
+        ! signed, endian, and scale_factor all have default values, so if they are
+        ! not present in the index file then set them as the default values, as reported
+        ! in section 3-53 of the WRF-AWR User's Guide
```
"AWR" -> "ARW"

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1025 @@
+        ! Calculate the range of this tile, which will be its filename
+        x(1) = (x(1) - modulo(x(1), tile_nx)) + 1
+        x(2) = x(1) + tile_nx
+        x(2) = x(2) - modulo(x(2), tile_nx)
```
I may be overlooking some special cases, but this logic seems more complicated than necessary. We know that `x(1)` is one more than a multiple of `tile_nx`, and by line 584, so is `x(2)`; so, couldn't we just subtract 1 from `x(2)`?

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1025 @@
+        ! NOTE: This calculation assumes that a dataset is a global dataset and may
+        ! not work correctly for non-global datasets
+        mgr % nTileX = int((360.0_RKIND / dx)) / tile_nx
+        mgr % nTileY = int((180.0_RKIND / dy)) / tile_ny
```
As with the calculation of `pixel_nx` and `pixel_ny`, I think we need to use `nint` here rather than `int`. Better still, we could consider just using `mgr % pixel_nx` and `mgr % pixel_ny` rather than duplicating the computation of those variables.

MiCurry replied:
Using `mgr % pixel_nx/ny` is a great idea!

MiCurry commented:
Okay @mgduda, I just pushed an update that addresses your above comments, except for changes on line 692 and 693, which I left a comment on!

MiCurry commented:
@mgduda I have just pushed an update to this branch which adds a default value to `tile_bdr` and uses `nint` on the `latlon_to_pixel` calculation and solves problems at the poles (as we discussed above). It also adds a new commit which adds a `mpas_geotile_manager.F` target to the `Makefile`.

mgduda commented:
It looks like the changes to `mpas_kd_tree.F` are being included in this PR as commit 04e88f0, while these were previously included in the `develop` branch as ddd137d . Could you rework the commit history of this PR so the tie-breaking commit doesn't appear as a new commit?

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1038 @@
+            pixel_x = pixel_x + mgr % pixel_nx
+        endif
+
+        if (pixel_y < 0.0) then
```
Since `pixel_y` is declared as an integer, should `0.0` be changed to `0`?

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1038 @@
+        endif
+
+        if (pixel_y < 0.0) then
+            pixel_y = pixel_y + 1
```
For negative `pixel_y` values, adding 1 doesn't seem like the correct thing to do. Is the general intent of this block of code to ensure that `pixel_y` is contained in the interval `[0, mgr %p ixel_ny - 1`]?

MiCurry replied:
@mgduda, yeah the intention was to contain pixel_y in the interval of `[0, mgr % pixel_ny - 1]`. Instead of adding 1 to negative values should we just set `pixel_y` to 0. Likewise with values of `pixel_y == mgr % pixel_nx - 1` instead of subtracting 1 should we set `pixel_y` to `mgr % pixel_nx - 1`?

Did you perhaps have a better solution in mind?

mgduda replied:
I agree that simply setting `pixel_y` to zero when its value is negative, and setting `pixel_y` to `mgr % pixel_ny - 1` when its value exceeds `mgr % pixel_ny - 1` would be the most robust solution.

MiCurry replied:
Sounds good

MiCurry commented:
@mgduda okay, I have just pushed an update that resolves rounding pixel_y values, changes 0.0 -> 0 and reworked the commit history to remove commit 04e88f0.

Let me know if you see anything else.

mgduda commented on src/core_init_atmosphere/mpas_geotile_manager.F:
```diff
@@ -0,0 +1,1038 @@
+        if (pixel_y < 0) then
+            pixel_y = 0
+        else if (pixel_y >= mgr % pixel_ny) then
+            pixel_y = pixel_y - 1
```
Should `pixel_y` be `mgr % pixel_ny` on the r.h.s.?

MiCurry replied:
I think you're right. I'll push an update.

MiCurry commented:
@mgduda Okay, I just pushed an update!
